sympy/solvers/solveset.py

```diff
@@ -871,3 +872,269 @@ def solveset(f, symbol=None):
+    This is done to maintain consistency in returning solutions
+    in the form of variable input by the user.
+
+    The algorithm used here is Gauss-Jordan elimination, which
+    results, after elimination, in an row echelon form matrix.
+
+    Returns
+    =======
+
+    A FiniteSet of ordered tuple of values of `symbols` for which
+    the `system` has solution.
+
+    Raises
+    ======
+
+    ValueError
+        The input is not valid.
+        The linear system has no solution.
+        The symbols are not given.
+
+    Examples
+    ========
+
+    >>> from sympy.solvers.solveset import linsolve
+    >>> from sympy import Matrix, S
+    >>> from sympy import symbols
+    >>> x, y, z = symbols("x, y, z")
+
+    >>> A = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 10]])
+    >>> b = Matrix([3, 6, 9])
+    >>> A
+    Matrix([
+    [1, 2,  3],
+    [4, 5,  6],
+    [7, 8, 10]])
+    >>> b
+    Matrix([
+    [3],
+    [6],
+    [9]])
+
+    >>> linsolve((A, b), [x, y, z])
+    {(-1, 2, 0)}
+
+    * Parametric Solution: In case the system is under determined, the function
+      will return parametric solution in terms of the given symbols.
+      Free symbols in the system are returned as it is. For e.g. in the system
+      below, `z` is returned as the solution for variable z, which means z is a
+      free symbol, i.e. it can take arbitrary values.
+
+    >>> A = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+    >>> b = Matrix([3, 6, 9])
+    >>> linsolve((A, b), [x, y, z])
+    {(z - 1, -2*z + 2, z)}
+
+    * List of Equations as input
+
+    >>> Eqns = [3*x + 2*y - z - 1, 2*x - 2*y + 4*z + 2, - x + S(1)/2*y - z]
+    >>> linsolve(Eqns, x, y, z)
+    {(1, -2, -2)}
+
+    * Augmented Matrix as input
+
+    >>> aug = Matrix([[2, 1, 3, 1], [2, 6, 8, 3], [6, 8, 18, 5]])
+    >>> aug
+    Matrix([
+    [2, 1,  3, 1],
+    [2, 6,  8, 3],
+    [6, 8, 18, 5]])
+    >>> linsolve(aug, x, y, z)
+    {(3/10, 2/5, 0)}
+
+    * Solve for symbolic coefficients
+
+    >>> a, b, c, d, e, f = symbols('a, b, c, d, e, f')
+    >>> eqns = [a*x + b*y - c, d*x + e*y - f]
+    >>> linsolve(eqns, x, y)
+    {(-b*(f - c*d/a)/(a*(e - b*d/a)) + c/a, (f - c*d/a)/(e - b*d/a))}
+
+    """
+
+    if not symbols:
+        raise ValueError('Symbols must be given, for which solution of the \
+                         system are to be found.')
+
+    if hasattr(symbols[0], '__iter__'):
+        symbols = symbols[0]
+
+    # 1). Augmented Matrix input Form
+    if isinstance(system, Matrix):
+        A, b = system[:, :-1], system[:, -1:]
+
+    elif hasattr(system, '__iter__'):
+
+        # 2). A & b as input Form
+        if len(system) == 2 and system[0].is_Matrix:
+            A, b = system[0], system[1]
+
+        # 3). List of equations Form
+        if not system[0].is_Matrix:
+            A, b = linear_eq_to_matrix(system, symbols)
+
+    else:
+        raise ValueError("Invalid arguments")
+
+    sol, params, free_syms = A.gauss_jordan_solve(b, parameters=True)
```

Comment:
This is mainly because, the time Complexity to examine A & b to determine which method to use  would be much higher than simply solving it through reduced row echleon form.

Moreover other methods can mainly solve square systems.